File: practiceApp/app/src/main/java/com/bomnie/practiceapp/MyAdapter.java
package com.bomnie.practiceapp;

import android.content.Context;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.facebook.drawee.backends.pipeline.Fresco;
import com.facebook.drawee.view.SimpleDraweeView;

import java.util.List;

public class MyAdapter extends RecyclerView.Adapter<MyAdapter.MyViewHolder> {
    private List<NewsData> mDataset;
    private static View.OnClickListener onClickListener;

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    public static class MyViewHolder extends RecyclerView.ViewHolder {
        // 항목화면에 들어가는 뷰 하나하나 연결
        // card_news.xml 안의 하나하나의 뷰 연결 - 계속해서 재사용 가능
        // each data item is just a string in this case
        public TextView tv_title;
        public TextView tv_content;
        public SimpleDraweeView iv_title;
        public View rootView;

        public MyViewHolder(View v) { // v: card_news.xml 의 root
            super(v);
            // ViewHolder = 메모리 관리와 성능 향상을 위해 안드로이드에서 친절히 만들어준 클래스
            // card_news.xml의 자식요소들을 재사용할 수 있도록
            tv_title = v.findViewById(R.id.tv_title);
            tv_content = v.findViewById(R.id.tv_content);
            iv_title = v.findViewById(R.id.iv_title);
            rootView = v;

            v.setClickable(true); // card_news.xml(LinearLayout)을 클릭 가능하게끔
            v.setEnabled(true);
            v.setOnClickListener(onClickListener);
            // 카드 하나의 어느 곳을 누르더라도 본문 내용으로 넘어가게 하도록 v에 설정
        }
    }

    // Provide a suitable constructor (depends on the kind of dataset)
    // mDataset: 각 card_news 뷰마다 보여줄 값을 가지고 있는 초기 데이터 원본
    public MyAdapter(List<NewsData> myDataset, Context context, View.OnClickListener onClick) {
        mDataset = myDataset;
        onClickListener = onClick;
        Fresco.initialize(context);

        // context는 Activity에서 받아오는 것이기 때문에 단순한 class에서는 사용가
    }

    // Create new views (invoked by the layout manager)
    @Override
    public MyAdapter.MyViewHolder onCreateViewHolder(ViewGroup parent,
                                                     int viewType) {
        // create a new view
        // 리사이클러 뷰의 항목화면 전체 연결
        // card_news.xml 전체 파일 연결
        LinearLayout v = (LinearLayout) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.card_news, parent, false);
        MyViewHolder vh = new MyViewHolder(v);
        return vh;
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(MyViewHolder holder, int position) {
        // - get element from your dataset at this position
        // - replace the contents of the view with that element
        NewsData news = mDataset.get(position);

        holder.tv_title.setText(news.getTitle());
        holder.tv_content.setText(news.getContent());

        Uri uri = Uri.parse(news.getUrlToImage());
        holder.iv_title.setImageURI(uri);

        // 여기에 클릭이벤트를 달면 갱신될 때마다 클릭이벤트를 생성하기 때문에 비효율적
        // tag - label
        holder.rootView.setTag(position); // v에 태그를 다는 작업

    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        return mDataset == null ? 0 : mDataset.size();
    }

    public NewsData getNews(int position){
        return mDataset != null ? mDataset.get(position) : null ;
    }
}

File: KotlinRecyclerViewApp/app/src/main/java/com/bomnie/kotlinrecyclerviewapp/MyAdapter.kt
package com.bomnie.kotlinrecyclerviewapp

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.Glide.init
import kotlinx.android.synthetic.main.recycler_item.view.*

class MyAdapter constructor(var context: Context, var items: ArrayList<Item>): RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        val inflater= LayoutInflater.from(context)
        val itemView= inflater.inflate(R.layout.recycler_item, parent, false)
        val vh= VH(itemView)

        return vh
    }

    override fun getItemCount(): Int {
        return items.size

    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val vh = holder as VH // 코틀린에서의 다운캐스팅 방법

        val item = items.get(position)
//        vh.tvTitle.setText(item.title) // 코틀린에서는 getter, setter 메소드를 권장하지 않음
        vh.tvTitle.text = item.title // text 를 멤버변수로 사용
        vh.tvMsg.text = item.msg

        // Glide 라이브러리 사용하여 이미지 설정 - 자바 코드와 동일
        Glide.with(context).load(item.img).into(vh.iv)

        // getLayoutPosition()을 사용할 수 없어서 여기에 클릭리스너를 추가
        vh.itemView.setOnClickListener(object: View.OnClickListener{
            override fun onClick(v: View?) {
                val intent= Intent(context, ItemActivity::class.java)
                intent.putExtra("title",item.title)
                intent.putExtra("msg",item.msg)
                intent.putExtra("image",item.img)

                context.startActivity(intent)
            }
        )


    // ViewHolder 클래스 (inner 클래스로 만들기)
    inner class VH constructor(itemView: View): RecyclerView.ViewHolder(itemView) {
        // 멤버변수[property]
        var tvTitle: TextView = itemView.tvTitle
        var tvMsg: TextView = itemView.tvMsg
        var iv: ImageView = itemView.iv
//        init{
//            tvTitle=itemView.findViewById(R.id.tvTitle)
//        }
    }
        init {
            // 자바에서는 이 곳에서 아이템 뷰 클릭 이벤트 처리를 했었음
            // 코틀린에서는 getLayoutPosition() 메소드의 결과가 언제나  -1만 나옴. 즉 포지션을 알 수 없음

            // 클릭리스너를 붙이는 작업을 onBindViewHolder()에서 작성
        }

    }
    }
}